zio/jsonio/encode.go

```diff
@@ -0,0 +1,245 @@
+		}
+		key := mapKey(typ.KeyType, it.Next())
+		e.entry(tab, key, zed.NewValue(typ.ValType, it.Next()))
+	}
+	e.newline()
+	e.indent(tab - e.tab)
+	e.punc('}')
+}
+
+func mapKey(typ zed.Type, b zcode.Bytes) string {
+	val := zed.NewValue(typ, b)
+	switch val.Type().Kind() {
+	case zed.PrimitiveKind:
+		if val.Type().ID() == zed.IDString {
+			// Don't quote strings.
+			return val.AsString()
+		}
+		return zson.FormatPrimitive(val.Type(), val.Bytes())
+	case zed.UnionKind:
+		// Untagged, decorated ZSON so
+		// |{0:1,0(uint64):2,0(=t):3,"0":4}| gets unique keys.
+		typ, bytes := typ.(*zed.TypeUnion).Untag(b)
+		return zson.FormatValue(zed.NewValue(typ, bytes))
+	case zed.EnumKind:
+		return convertEnum(typ.(*zed.TypeEnum), b)
+	default:
+		return zson.FormatValue(val)
+	}
+}
+
+func (e *encoder) marshalEnum(typ *zed.TypeEnum, bytes zcode.Bytes) {
+	e.writeColor(e.marshalJSON(convertEnum(typ, bytes)), stringColor)
+}
+
+func convertEnum(typ *zed.TypeEnum, bytes zcode.Bytes) string {
+	if k := int(zed.DecodeUint(bytes)); k < len(typ.Symbols) {
+		return typ.Symbols[k]
+	}
+	return "<bad enum>"
+}
+
+func (e *encoder) marshalError(tab int, typ *zed.TypeError, bytes zcode.Bytes) {
+	tab += e.tab
+	e.punc('{')
+	e.entry(tab, "error", zed.NewValue(typ.Type, bytes))
+	e.newline()
+	e.indent(tab - e.tab)
+	e.punc('}')
+}
+
+func (e *encoder) entry(tab int, name string, val zed.Value) {
+	e.newline()
+	e.indent(tab)
+	e.writeColor(e.marshalJSON(name), fieldColor)
+	e.punc(':')
+	if e.tab != 0 {
+		e.buf.WriteByte(' ')
+	}
+	e.marshal(tab, val)
+}
+
+func (e *encoder) marshalPrimitive(val zed.Value) {
+	var v any
+	c := stringColor
+	switch id := val.Type().ID(); {
+	case id == zed.IDDuration:
+		v = nano.Duration(val.Int()).String()
+	case id == zed.IDTime:
+		v = nano.Ts(val.Int()).Time().Format(time.RFC3339Nano)
+	case zed.IsSigned(id):
+		v, c = val.Int(), numberColor
+	case zed.IsUnsigned(id):
+		v, c = val.Uint(), numberColor
+	case zed.IsFloat(id):
+		v, c = val.Float(), numberColor
+	case id == zed.IDBool:
+		v, c = val.AsBool(), boolColor
+	case id == zed.IDBytes:
+		v = "0x" + hex.EncodeToString(val.Bytes())
+	case id == zed.IDString:
+		v = val.AsString()
+	case id == zed.IDIP:
+		v = zed.DecodeIP(val.Bytes()).String()
+	case id == zed.IDNet:
+		v = zed.DecodeNet(val.Bytes()).String()
+	case id == zed.IDType:
+		v = zson.FormatValue(val)
+	default:
+		v = fmt.Sprintf("<unsupported id=%d>", id)
+	}
+	e.writeColor(e.marshalJSON(v), c)
+}
+
+func (e *encoder) marshalJSON(v any) []byte {
+	e.primBuf.Reset()
+	if err := e.primEnc.Encode(v); err != nil {
+		panic(err)
+	}
+	return bytes.TrimSpace(e.primBuf.Bytes())
+}
+
+func (e *encoder) punc(b byte) {
+	e.writeColor([]byte{b}, puncColor)
+}
+
+func (e *encoder) writeColor(b []byte, code []byte) {
+	if e.tab > 0 && color.Enabled {
+		e.buf.Write(code)
+		defer e.buf.WriteString(color.Reset.String())
+	}
+	e.buf.Write(b)
+}
+
+func (e *encoder) newline() {
+	e.buf.WriteString(e.newlineChar)
```

Comment:
Nit: No need for newlineChar if you do this.
```suggestion
	if e.tab > 0 {
		e.buf.WriteByte(`\n`)
	}
```